### ISSUE TITLE
Clustered webapp race condition writing [] to memcache

### ISSUE THREAD

**Issue by apg-pk** [graphite-project/graphite-web#1278]
Using Graphite 0.9.13 in a 3 node cluster, on initial cache miss for a render request, the correct datapoints are returned, but subsequent requests return an empty array []:

```bash
$ curl "0/render/?salt=1436735833.313&target=servers.cmgm004.GenericJMX-QPID-CLIENT-consumerCount.gauge-Count&format=json&from=-1min"
[{"target": "servers.cmgm004.GenericJMX-QPID-CLIENT-consumerCount.gauge-Count", "datapoints": [[12.0, 1436767950], [12.0, 1436767960], [12.0, 1436767970], [12.0, 1436767980], [12.0, 1436767990], [null, 1436768000]]}]

$ curl "0/render/?salt=1436735833.313&target=servers.cmgm004.GenericJMX-QPID-CLIENT-consumerCount.gauge-Count&format=json&from=-1min"
[]

$ curl "0/render/?salt=1436735833.313&target=servers.cmgm004.GenericJMX-QPID-CLIENT-consumerCount.gauge-Count&format=json&from=-1min"
[]
```

Using noCache query param or turning off memcached appear to be workarounds.

Problem appears to be in: https://github.com/graphite-project/graphite-web/blob/0.9.13-pre1/webapp/graphite/render/views.py#L126-L127

```python
     for target in targets:
        t = time()
        seriesList = evaluateTarget(requestContext, target)
        log.rendering("Retrieval of %s took %.6f" % (target, time() - t))
        data.extend(seriesList)

      if useCache:
        cache.add(dataKey, data, cacheTimeout)
```

If **data** is never populated for any of the targets (because no metrics are on current node), then **data** will stay initialized to **[]** and then added to cache for subsequent requests to get.

**Comment by deniszh** [comment]
Are you using shared memcached? It's known issue then. Please switch to local memcached instance for every node.

**Comment by deniszh** [comment]
Adding empty list to cache is harmless btw, just turn off shared memcached.

**Comment by apg-pk** [comment]
Yes, shared memcached, but unfortunately not harmless. The nodes without the metric are able to insert an empty [] into the cache. The empty list is considered a cache hit  and sent in the response even though the cache log reports a miss. In the code below, the empty list is considered **false** for the conditional but also **not None** for the actual request context:

```python
    if useCache:
      targets = requestOptions['targets']
      startTime = requestOptions['startTime']
      endTime = requestOptions['endTime']
      dataKey = hashData(targets, startTime, endTime)
      cachedData = cache.get(dataKey)
      if cachedData:
        log.cache("Data-Cache hit [%s]" % dataKey)
      else:
        log.cache("Data-Cache miss [%s]" % dataKey)
    else:
      cachedData = None

    if cachedData is not None:
      requestContext['data'] = data = cachedData
    else: # Have to actually retrieve the data now
```

https://github.com/graphite-project/graphite-web/blob/0.9.13-pre1/webapp/graphite/render/views.py#L98-L113

**Comment by deniszh** [comment]
Not sure if that's enough to fix shared memcached problem. IMO PR #1212 is more general fix for this.

**Comment by apg-pk** [comment]
Hrm... we are using REPLICATION_FACTOR=1 and getting empty metrics, so perhaps this is a slightly different issue from getting redundant metric time series from clustered backends w/ REPLICATION_FACTOR>=2?  PR #1279 just prevents empty results from getting into shared memcached and locking out the actual results.

**Comment by deniszh** [comment]
I also faced different issues with shared memcached and replication factor = 1. My point is that #1279 will add hostname to caching key, so, it will mitigate this issue too, probably. But it's just my hypothesis, didn't test that yet.

**Comment by obfuscurity** [comment]
Closing this since the proposed fix in #1279 has been merged and we deal more explicitly with the shared memcached issue elsewhere, e.g. #1212.
